site/gatsby-site/migrations/2023.04.17T23.15.19.remove-cset-publish-field.js

```diff
@@ -0,0 +1,56 @@
+const config = require('../config');
+
+/** @type {import('umzug').MigrationFn<any>} */
+exports.up = async ({ context: { client } }) => {
+  // We're getting duplicate publish fields because
+  // publish was a field of the CSET taxonomy,
+  // but it has been upgraded to a top-level member of all taxonomies.
+  // To remove the duplicate we need to delete the original publish field
+  // from the CSET taxonomy and use only the new method.
+  const taxaCollection = client.db(config.realm.production_db.db_name).collection('taxa');
+
+  const csetTaxonomy = await taxaCollection.findOne({ namespace: 'CSET' });
+
+  csetTaxonomy.field_list = csetTaxonomy.field_list.filter((f) => f.short_name != 'Publish');
+
+  delete csetTaxonomy._id;
+
+  await taxaCollection.updateOne({ namespace: 'CSET' }, { $set: { ...csetTaxonomy } });
+
+  // Then we need to remove all the publish attributes from the classifications,
+  // ensuring that the publish status remains the same.
+  const classificationsCollection = client
+    .db(config.realm.production_db.db_name)
+    .collection('classifications');
+
+  const classifications = await classificationsCollection.find({ namespace: 'CSET' });
+
+  while (await classifications.hasNext()) {
+    const classification = await classifications.next();
+
+    const attributePublish = classification.attributes.find(
+      (attribute) => attribute.short_name == 'Publish'
+    );
+
+    const attributePublishValue =
+      attributePublish?.value_json && JSON.parse(attributePublish.value_json);
+
+    // Ensure that we don't have a situation where one is explicitly set to true
+    // and the other explicitly set to false.
+    console.assert(Boolean(classification.publish) == Boolean(attributePublishValue));
```

Comment:
To be precise, this doesn't ensure what you describe. It only prints an error message, but the next sentences still execute. I don't know if that's what you want here.

Reply:
I don't think this condition will actually be met, but I changed it to throw an exception instead.